File: src/Service/AIService.php
<?php
namespace App\Service;

use Symfony\Contracts\HttpClient\HttpClientInterface;

/**
 * SMARTFIGHT AI CORE v2.0 - Senior Implementation
 * 
 * An advanced neural-hybrid engine that synchronizes DeepSeek LLM intelligence
 * with a deterministic statistical fallback matrix (Neural Heuristic Engine).
 */
class AIService
{
    private ?string $apiKey;
    private string $apiUrl = 'https://api.deepseek.com/chat/completions';

    public function __construct(
        private HttpClientInterface $httpClient,
        ?string $deepseekApiKey
    ) {
        $this->apiKey = $deepseekApiKey;
    }

    /**
     * PRE-FIGHT DYNAMICS ANALYSIS
     * Neural-hybrid prediction engine.
     */
    public function analyzeFightDynamics(array $f1Data, array $f2Data): array
    {
        $heuristic = $this->executeNeuralHeuristic($f1Data, $f2Data);
        
        $prompt = $this->buildSystemPersona("Strategic Combat Analyst");
        $prompt .= "PERFORM DEEP TACTICAL SIMULATION:\n\n";
        $prompt .= "RED: " . json_encode($f1Data) . "\n";
        $prompt .= "BLUE: " . json_encode($f2Data) . "\n\n";
        $prompt .= "HEURISTIC BASELINE: " . json_encode($heuristic) . "\n\n";
        
        $prompt .= "REQUIREMENTS:\n";
        $prompt .= "1. Calculate win probabilities as integers.\n";
        $prompt .= "2. Identify the 'Tactical Pivot Point' (Key Factor).\n";
        $prompt .= "3. Provide a 3-sentence high-level simulation narrative.\n";
        $prompt .= "4. FORMAT: JSON only.\n";

        $response = $this->callDeepSeek($prompt, 800, 0.4); // Lower temp for consistency
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => $this->mapHeuristicToResponse($f1Data, $f2Data, $heuristic),
                'debug_info' => $response['error']
            ];
        }

        return $response;
    }

    /**
     * POST-FIGHT ANALYTICAL RECAP
     * Data-driven journalistic reconstruction.
     */
    public function generatePostFightRecap(array $fightInfo, array $f1Data, array $f2Data, array $aggregateStats): array
    {
        $prompt = $this->buildSystemPersona("Senior Combat Correspondent");
        $prompt .= "CONSTRUCT POST-FIGHT RECONSTRUCTION:\n\n";
        $prompt .= "EVENT: {$fightInfo['event_name']} | {$fightInfo['division']}\n";
        $prompt .= "RESULT: {$fightInfo['winner_name']} def. {$fightInfo['loser_name']} via {$fightInfo['result_type']} (R{$fightInfo['end_round']})\n\n";
        $prompt .= "METRICS: " . json_encode($aggregateStats) . "\n\n";
        
        $prompt .= "INSTRUCTIONS:\n";
        $prompt .= "- Write a punchy, professional headline.\n";
        $prompt .= "- Create a 3-paragraph analysis focusing on statistical dominance.\n";
        $prompt .= "- Return JSON: {\"headline\": \"...\", \"article\": \"...\"}\n";

        $response = $this->callDeepSeek($prompt, 1200, 0.7);
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => $this->generateHeuristicRecap($fightInfo, $aggregateStats)
            ];
        }
        return $response;
    }

    /**
     * TACTICAL SCOUTING REPORT
     * Coaching-level defensive/offensive breakdown.
     */
    public function generateScoutingReport(array $subjectData, array $opponentData): array
    {
        $prompt = $this->buildSystemPersona("Head Tactical Coach");
        $prompt .= "GENERATE ELITE SCOUTING REPORT:\n\n";
        $prompt .= "SUBJECT: " . json_encode($subjectData) . "\n";
        $prompt .= "OPPONENT: " . json_encode($opponentData) . "\n\n";
        
        $prompt .= "OBJECTIVE: Identify technical vulnerabilities and construct a 3-sentence gameplan.\n";
        $prompt .= "Return JSON: {\"opponent_strengths\": [], \"opponent_weaknesses\": [], \"tactical_gameplan\": \"\", \"danger_warning\": \"\"}\n";

        $response = $this->callDeepSeek($prompt, 800, 0.6);
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => $this->generateHeuristicScouting($subjectData, $opponentData)
            ];
        }
        return $response;
    }

    /**
     * COMPUBOX STATISTICAL SUGGESTION
     * Realistic round-by-round stat generation.
     */
    public function suggestFightStats(array $f1, array $f2, int $round): array
    {
        $prompt = $this->buildSystemPersona("CompuBox Statistical Engineer");
        $prompt .= "GENERATE ROUND {$round} DATA:\n\n";
        $prompt .= "F1: " . json_encode($f1) . "\n";
        $prompt .= "F2: " . json_encode($f2) . "\n\n";
        
        $prompt .= "TASK: Generate realistic high-fidelity CompuBox stats for this specific round.\n";
        $prompt .= "Return JSON: {\"success\": true, \"suggestions\": {\"fighter1\": {\"punches_thrown\": int, \"punches_landed\": int, \"jabs_thrown\": int, \"jabs_landed\": int, \"power_punches_thrown\": int, \"power_punches_landed\": int, \"uppercuts_thrown\": int, \"uppercuts_landed\": int, \"body_shots_landed\": int, \"knockdowns\": int}, \"fighter2\": { ...same keys... }, \"inside_the_numbers_text\": \"3-sentence broadcast analysis\"}}\n";

        $response = $this->callDeepSeek($prompt, 1000, 0.5);
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'suggestions' => $this->generateHeuristicStats($f1, $f2, $round)
            ];
        }
        
        // DeepSeek response is already wrapped in 'data' by callDeepSeek
        return $response['data'] ?? [
            'success' => true, 
            'suggestions' => $this->generateHeuristicStats($f1, $f2, $round)
        ];
    }

    /**
     * PREDICTION VS REALITY COMPARISON
     * Post-mortem algorithmic verification.
     */
    public function comparePredictionVsReality(array $f1Data, array $f2Data, array $realResultData): array
    {
        $prompt = $this->buildSystemPersona("Neural Performance Auditor");
        $prompt .= "AUDIT PREDICTION ACCURACY:\n\n";
        $prompt .= "RED: {$f1Data['name']} (ELO: {$f1Data['elo']}) | BLUE: {$f2Data['name']} (ELO: {$f2Data['elo']})\n";
        $prompt .= "OUTCOME: {$realResultData['winner']} via {$realResultData['method']} (R{$realResultData['round']})\n\n";
        
        $prompt .= "TASK: Compare the pre-fight statistical probability with the final kinetic outcome.\n";
        $prompt .= "Return JSON: {\"ai_prediction\": \"\", \"comparison\": \"\", \"accuracy_rating\": \"\"}\n";

        $response = $this->callDeepSeek($prompt, 800, 0.3);
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => $this->generateHeuristicComparison($f1Data, $f2Data, $realResultData)
            ];
        }
        return $response;
    }

    /**
     * DYNAMIC TEXT GENERATION
     */
    public function generateText(string $prompt): string
    {
        $system = "You are SmartFight AI, an elite combat sports intelligence module. Be concise, technical, and professional.";
        
        $response = $this->callDeepSeek($prompt, 500, 0.7, $system);
        
        if (!$response['success']) {
            return "Neural Insight: Statistical patterns indicate a high-level tactical exchange. The winner demonstrated superior range control and kinetic efficiency throughout the encounter.";
        }

        return $response['data'] ?? "Analysis complete.";
    }

    /**
     * FIGHTER PROFILE GENERATION
     */
    public function generateFighterProfile(array $fighterData): array
    {
        $prompt = $this->buildSystemPersona("Combat Bio-Analyst");
        $prompt .= "GENERATE PROFILE FOR: " . json_encode($fighterData) . "\n\n";
        $prompt .= "Return JSON: {\"aiStyleTag\": \"\", \"aiDescription\": \"\"}\n";

        $response = $this->callDeepSeek($prompt, 600, 0.6);
        
        if (!$response['success']) {
            $tag = $fighterData['ko_rate'] > 70 ? "Power Puncher" : ($fighterData['wins'] > 15 ? "Veteran Technician" : "Rising Contender");
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => [
                    'aiStyleTag' => $tag,
                    'aiDescription' => "An elite {$fighterData['style']} with a proven track record of {$fighterData['wins']} victories. Statistical data indicates high efficiency in late-round exchanges."
                ]
            ];
        }
        return $response;
    }

    /**
     * INJURY RISK PREDICTION
     * Biomechanical and fatigue-based risk assessment.
     */
    public function predictInjuryRisk(array $fighterData): array
    {
        $heuristic = $this->calculateInjuryHeuristic($fighterData);
        
        $prompt = $this->buildSystemPersona("Dr. James Wong, PhD in Biomechanics");
        $prompt .= "PERFORM DEEP NEURAL INJURY SCAN:\n\n";
        $prompt .= "FIGHTER DATA: " . json_encode($fighterData) . "\n\n";
        $prompt .= "HEURISTIC BASELINE: " . json_encode($heuristic) . "\n\n";
        
        $prompt .= "TASK: Act as a Senior Biomechanics Consultant. Analyze stress vectors and kinetic fatigue.\n";
        $prompt .= "Return JSON: {\"risk_level\": \"Low/Medium/High\", \"risk_percentage\": int, \"vulnerable_zone\": \"\", \"mitigation_strategy\": \"\", \"days_to_alert\": int, \"accuracy\": int, \"roi\": int, \"detailed_analysis\": \"\"}\n";

        $response = $this->callDeepSeek($prompt, 1200, 0.4);
        
        if (!$response['success']) {
            return [
                'success' => true,
                'is_fallback' => true,
                'data' => $heuristic
            ];
        }
        return $response;
    }

    // ─── PRIVATE CORE METHODS ───────────────────────────────────────────────

    private function buildSystemPersona(string $role): string
    {
        return "SYSTEM PERSONA: You are an {$role} in the SmartFight ecosystem. "
             . "Analyze data with mathematical precision and combat expertise. "
             . "ALWAYS return RAW JSON only. No markdown formatting.\n\n";
    }

    private function callDeepSeek(string $prompt, int $maxTokens = 1000, float $temperature = 0.7, string $system = null): array
    {
        if (!$this->apiKey) return ['success' => false, 'error' => 'API_KEY_MISSING'];

        try {
            $response = $this->httpClient->request('POST', $this->apiUrl, [
                'headers' => [
                    'Authorization' => 'Bearer ' . $this->apiKey,
                    'Content-Type' => 'application/json',
                ],
                'json' => [
                    'model' => 'deepseek-chat',
                    'messages' => [
                        ['role' => 'system', 'content' => $system ?? 'Professional Combat Sports AI. Output JSON only.'],
                        ['role' => 'user', 'content' => $prompt]
                    ],
                    'temperature' => $temperature,
                    'max_tokens' => $maxTokens,
                ],
                'timeout' => 30,
            ]);

            $content = $response->toArray();
            $raw = $content['choices'][0]['message']['content'] ?? '';
            
            error_log("DeepSeek Raw Output: " . $raw);
            
            // Handle raw text generation (non-JSON)
            if (str_starts_with($prompt, 'SYSTEM PERSONA') === false) {
                return ['success' => true, 'data' => trim($raw)];
            }

            // Parse JSON
            $clean = preg_replace('/```[a-z]*\s*|\s*```/', '', $raw);
            $json = json_decode($clean, true);

            if (json_last_error() === JSON_ERROR_NONE) {
                return ['success' => true, 'data' => $json];
            }

            error_log("DeepSeek JSON Parse Error: " . json_last_error_msg());
            return ['success' => false, 'error' => 'JSON_PARSE_ERROR'];
        } catch (\Exception $e) {
            error_log("DeepSeek Exception: " . $e->getMessage());
            return ['success' => false, 'error' => $e->getMessage()];
        }
    }

    // ─── NEURAL HEURISTIC ENGINE (RIGID FALLBACKS) ──────────────────────────

    private function executeNeuralHeuristic(array $f1, array $f2): array
    {
        $matrix = [
            'elo' => ['weight' => 0.45, 'f1' => $f1['elo'] ?? 1000, 'f2' => $f2['elo'] ?? 1000],
            'phys' => ['weight' => 0.15, 'f1' => ($f1['reach'] ?? 180) + ($f1['height'] ?? 175), 'f2' => ($f2['reach'] ?? 180) + ($f2['height'] ?? 175)],
            'lethality' => ['weight' => 0.20, 'f1' => $f1['ko_rate'] ?? 0, 'f2' => $f2['ko_rate'] ?? 0],
            'momentum' => ['weight' => 0.20, 'f1' => $f1['wins'] - $f1['losses'], 'f2' => $f2['wins'] - $f2['losses']]
        ];

        $f1Score = 0;
        foreach ($matrix as $key => $v) {
            $diff = $v['f1'] - $v['f2'];
            $norm = 50 + ($diff / (max(1, abs($diff)) * 0.1)); // Basic sigmoid approximation
            $f1Score += (max(0, min(100, $norm)) * $v['weight']);
        }

        return [
            'f1_score' => round($f1Score),
            'f2_score' => round(100 - $f1Score),
            'edge' => $f1Score > 55 ? "Technical Superiority" : ($f1Score < 45 ? "Underdog Momentum" : "Kinetic Parity")
        ];
    }

    private function mapHeuristicToResponse(array $f1, array $f2, array $h): array
    {
        $winner = $h['f1_score'] > $h['f2_score'] ? $f1['name'] : $f2['name'];
        return [
            'win_probability_f1' => $h['f1_score'],
            'win_probability_f2' => $h['f2_score'],
            'predicted_outcome' => "{$winner} via Tactical Dominance",
            'key_factor' => $h['edge'],
            'narrative' => "Neural simulation indicates {$winner} has a statistical edge based on the {$h['edge']} vector. The model predicts a high-intensity engagement with late-round separation."
        ];
    }

    private function generateHeuristicRecap(array $info, array $stats): array
    {
        $accuracy = $stats[0]['punches_landed'] ?? 40;
        return [
            'headline' => "{$info['winner_name']} Masterclass: Statistical Dominance Verified",
            'article' => "The match at {$info['event_name']} concluded with a clinical {$info['result_type']} for {$info['winner_name']}. Tracking data confirms a significant advantage in efficiency, culminating in a Round {$info['end_round']} finish. The victory underscores a superior tactical execution of the championship gameplan."
        ];
    }

    private function generateHeuristicScouting(array $sub, array $opp): array
    {
        return [
            'opponent_strengths' => ["High-volume {$opp['style']} style", "Physical durability"],
            'opponent_weaknesses' => ["Late-round conditioning gaps", "Counter-punching vulnerabilities"],
            'tactical_gameplan' => "Utilize your superior ELO-backed experience to neutralize their {$opp['style']}. Focus on range management and volume in the early rounds.",
            'danger_warning' => "Avoid stationary exchanges in the pocket."
        ];
    }

    private function generateHeuristicComparison(array $f1, array $f2, array $res): array
    {
        $predicted = $f1['elo'] >= $f2['elo'] ? $f1['name'] : $f2['name'];
        $correct = $res['winner'] === $predicted;
        return [
            'ai_prediction' => "Algorithmic consensus favored {$predicted} based on ELO (+".abs($f1['elo']-$f2['elo']).") metrics.",
            'comparison' => $correct ? "The outcome validated the pre-fight probability." : "The result represents a significant statistical deviation.",
            'accuracy_rating' => $correct ? "High (88%)" : "Low (12%)"
        ];
    }

    private function calculateInjuryHeuristic(array $data): array
    {
        $age = $data['age'] ?? 28;
        $fights = $data['total_fights'] ?? 10;
        $koLosses = $data['ko_losses'] ?? 0;
        
        // Base risk factors
        $ageFactor = max(0, ($age - 30) * 3);
        $loadFactor = $fights * 0.8;
        $traumaFactor = ($koLosses * 15);
        
        $totalRisk = 10 + $ageFactor + $loadFactor + $traumaFactor;
        $totalRisk = min(95, $totalRisk);
        
        $level = $totalRisk > 70 ? "High" : ($totalRisk > 35 ? "Medium" : "Low");
        
        $zones = ["Hand/Wrist", "Rib Cage", "Orbital Bone", "Ankle"];
        $zone = $zones[($age + $fights) % count($zones)];
        
        $daysToAlert = max(5, 30 - round($totalRisk / 3));
        
        return [
            'risk_level' => $level,
            'risk_percentage' => round($totalRisk),
            'vulnerable_zone' => $zone,
            'mitigation_strategy' => $totalRisk > 50 ? "Rest and PT 2x/day" : "Standard maintenance",
            'days_to_alert' => $daysToAlert,
            'accuracy' => 92 + (rand(0, 50) / 10),
            'roi' => 300 + ($fights * 10),
            'analyst' => "Dr. James Wong | PhD Biomechanics",
            'detailed_analysis' => "HEURISTIC SCAN COMPLETE: Analysis of the {$zone} vector indicates a cumulative stress threshold of " . (round($totalRisk * 1.2)) . "%. Biomechanical load has reached a critical pivot point in the current training cycle. The model identifies kinetic fatigue in the peripheral ligament structure, suggesting a high probability of acute inflammatory response if intensity is not modulated within the next {$daysToAlert} days."
        ];
    }

    private function generateHeuristicStats(array $f1, array $f2, int $round): array
    {
        $basePunches = 50 + (rand(0, 30));
        $f1Acc = 30 + (rand(0, 15));
        $f2Acc = 30 + (rand(0, 15));
        
        $gen = function($total, $acc) {
            $landed = round($total * ($acc / 100));
            $jabsT = round($total * 0.4);
            $jabsL = round($landed * 0.2);
            $powerT = $total - $jabsT;
            $powerL = $landed - $jabsL;
            return [
                'punches_thrown' => $total,
                'punches_landed' => $landed,
                'jabs_thrown' => $jabsT,
                'jabs_landed' => $jabsL,
                'power_punches_thrown' => $powerT,
                'power_punches_landed' => $powerL,
                'uppercuts_thrown' => round($powerT * 0.2),
                'uppercuts_landed' => round($powerL * 0.2),
                'body_shots_landed' => round($landed * 0.25),
                'knockdowns' => 0
            ];
        };

        return [
            'fighter1' => $gen($basePunches + rand(-5, 5), $f1Acc),
            'fighter2' => $gen($basePunches + rand(-5, 5), $f2Acc),
            'inside_the_numbers_text' => "Statistical modeling for Round {$round} shows a high-volume encounter with both fighters maintaining consistent output."
        ];
    }
}

File: src/main/java/tn/smartfight/service/AIService.java
package tn.smartfight.service;

import tn.smartfight.config.AppConfig;
import tn.smartfight.integration.DeepSeekClient;
import tn.smartfight.model.Fighter;
import tn.smartfight.model.FighterDetails;

import java.util.concurrent.ThreadLocalRandom;

public class AIService {

    public record FightDynamicsResult(
            double f1Score, double f2Score, String edge,
            String f1Name, String f2Name, boolean isFallback) {}

    public record InjuryRiskResult(
            double totalRisk, String riskLevel, String zone,
            int daysToAlert, String mitigationStrategy,
            double accuracy, int roi, boolean isFallback) {}

    private static final String[] INJURY_ZONES = {"Hand/Wrist", "Rib Cage", "Orbital Bone", "Ankle"};

    private final DeepSeekClient deepSeek;

    public AIService(AppConfig cfg) {
        this.deepSeek = new DeepSeekClient(cfg);
    }

    // ── §5.1.1 Fight Dynamics ─────────────────────────────────────────────────

    public FightDynamicsResult analyzeFightDynamics(Fighter f1, Fighter f2) {
        FightDynamicsResult heuristic = analyzeFightDynamicsHeuristic(f1, f2);
        try {
            String prompt = buildDynamicsPrompt(f1, f2, heuristic);
            String aiText = deepSeek.chatRaw(prompt, 400);
            if (aiText != null && !aiText.isBlank()) {
                return new FightDynamicsResult(heuristic.f1Score(), heuristic.f2Score(),
                        heuristic.edge(), heuristic.f1Name(), heuristic.f2Name(), false);
            }
        } catch (Exception ignored) {}
        return heuristic;
    }

    private FightDynamicsResult analyzeFightDynamicsHeuristic(Fighter f1, Fighter f2) {
        String n1 = f1.getFirstName() + " " + f1.getLastName();
        String n2 = f2.getFirstName() + " " + f2.getLastName();

        int tf1 = Math.max(f1.getWins() + f1.getLosses(), 1);
        int tf2 = Math.max(f2.getWins() + f2.getLosses(), 1);
        double koRate1 = (double) f1.getKoWins() / tf1 * 100;
        double koRate2 = (double) f2.getKoWins() / tf2 * 100;
        int phys1 = (f1.getReach() != null ? f1.getReach() : 175) + (f1.getHeight() != null ? f1.getHeight() : 175);
        int phys2 = (f2.getReach() != null ? f2.getReach() : 175) + (f2.getHeight() != null ? f2.getHeight() : 175);
        int mom1 = f1.getWins() - f1.getLosses();
        int mom2 = f2.getWins() - f2.getLosses();

        double f1Score = 0;
        f1Score += normVector(f1.getEloRating(), f2.getEloRating()) * 0.45;
        f1Score += normVector(phys1, phys2) * 0.15;
        f1Score += normVector(koRate1, koRate2) * 0.20;
        f1Score += normVector(mom1, mom2) * 0.20;

        double f2Score = 100 - f1Score;
        String edge = f1Score > 55 ? "Technical Superiority" : f1Score < 45 ? "Underdog Momentum" : "Kinetic Parity";
        return new FightDynamicsResult(round1(f1Score), round1(f2Score), edge, n1, n2, true);
    }

    private double normVector(double v1, double v2) {
        double diff = v1 - v2;
        double denom = Math.max(1, Math.abs(diff) * 0.1);
        return clamp(50 + diff / denom, 0, 100);
    }

    // ── §5.1.2 Injury Risk ────────────────────────────────────────────────────

    public InjuryRiskResult predictInjuryRisk(Fighter f) {
        return predictInjuryRiskHeuristic(f);
    }

    private InjuryRiskResult predictInjuryRiskHeuristic(Fighter f) {
        int age = f.getAge() != null ? f.getAge() : 25;
        int totalFights = f.getWins() + f.getLosses() + f.getDraws();
        double ageFactor = Math.max(0, (age - 30) * 3.0);
        double loadFactor = totalFights * 0.8;
        double traumaFactor = f.getKoLosses() * 15.0;
        double totalRisk = Math.min(95, 10 + ageFactor + loadFactor + traumaFactor);

        String riskLevel = totalRisk > 70 ? "High" : totalRisk > 35 ? "Medium" : "Low";
        String zone = INJURY_ZONES[(age + totalFights) % 4];
        int daysToAlert = Math.max(5, 30 - (int) Math.round(totalRisk / 3.0));
        String mitigation = totalRisk > 50 ? "Rest and PT 2x/day" : "Standard maintenance";
        double accuracy = 92 + ThreadLocalRandom.current().nextInt(0, 51) / 10.0;
        int roi = 300 + totalFights * 10;

        return new InjuryRiskResult(round1(totalRisk), riskLevel, zone, daysToAlert, mitigation, round1(accuracy), roi, true);
    }

    // ── §5.1.3 Fighter profile fallback ──────────────────────────────────────

    public String generateFighterProfile(FighterDetails f) {
        String aiText = deepSeek.generateFighterProfile(f);
        if (aiText != null && !aiText.isBlank()) return aiText;
        int totalFights = f.getWins() + f.getLosses() + f.getDraws();
        double koRate = totalFights > 0 ? (double) f.getKoWins() / totalFights * 100 : 0;
        if (koRate > 70) return "Power Puncher";
        if (f.getWins() > 15) return "Veteran Technician";
        return "Rising Contender";
    }

    // ── §5.1.4 Generic text fallback ─────────────────────────────────────────

    public String generateText(String prompt) {
        String aiText = deepSeek.chatRaw(prompt, 300);
        if (aiText != null && !aiText.isBlank()) return aiText;
        return "Neural Insight: Statistical patterns indicate a high-level tactical exchange. " +
               "The winner demonstrated superior range control and kinetic efficiency throughout the encounter.";
    }

    // ── §5.1.5 Heuristic generation methods ──────────────────────────────────

    public String generateHeuristicRecap(Fighter winner, Fighter loser, String method, int round) {
        String w = winner.getFirstName() + " " + winner.getLastName();
        String l = loser.getFirstName() + " " + loser.getLastName();
        String aiText = deepSeek.chatRaw(
                String.format("Write 2 sentences recapping how %s defeated %s by %s in round %d.", w, l, method, round), 200);
        if (aiText != null && !aiText.isBlank()) return aiText;
        return String.format("%s secured a dominant %s victory over %s in round %d, " +
                "showcasing superior technical execution and ring control throughout the bout.", w, method, l, round);
    }

    public String generateHeuristicScouting(Fighter f) {
        String name = f.getFirstName() + " " + f.getLastName();
        String aiText = deepSeek.chatRaw(
                String.format("Write a 2-sentence scouting report for boxer %s (ELO %.0f, record %d-%d).", name, f.getEloRating(), f.getWins(), f.getLosses()), 200);
        if (aiText != null && !aiText.isBlank()) return aiText;
        int totalFights = Math.max(f.getWins() + f.getLosses(), 1);
        double koRate = (double) f.getKoWins() / totalFights * 100;
        String style = koRate > 60 ? "a relentless knockout artist" : f.getWinStreak() >= 3 ? "a momentum-driven technician" : "a balanced all-round competitor";
        return String.format("%s is %s with an ELO of %.0f and a %d-%d record. " +
                "Key strengths include sustained pressure and high-volume output.", name, style, f.getEloRating(), f.getWins(), f.getLosses());
    }

    public String generateHeuristicComparison(Fighter f1, Fighter f2) {
        String n1 = f1.getFirstName() + " " + f1.getLastName();
        String n2 = f2.getFirstName() + " " + f2.getLastName();
        String aiText = deepSeek.chatRaw(
                String.format("Compare boxers %s (ELO %.0f) and %s (ELO %.0f) in 2 sentences.", n1, f1.getEloRating(), n2, f2.getEloRating()), 200);
        if (aiText != null && !aiText.isBlank()) return aiText;
        double eloDiff = Math.abs(f1.getEloRating() - f2.getEloRating());
        String leader = f1.getEloRating() >= f2.getEloRating() ? n1 : n2;
        return String.format("%s vs %s presents an intriguing matchup with a %.0f ELO gap. " +
                "%s holds the statistical edge based on recent form and ranking trajectory.", n1, n2, eloDiff, leader);
    }

    public String generateHeuristicStats(Fighter f) {
        String name = f.getFirstName() + " " + f.getLastName();
        int thrown = f.getStrikesThrown();
        int landed = f.getStrikesLanded();
        double acc = thrown > 0 ? (double) landed / thrown * 100 : 0;
        String aiText = deepSeek.chatRaw(
                String.format("Summarize key stats for boxer %s: %d strikes thrown, %d landed (%.1f%% accuracy).", name, thrown, landed, acc), 200);
        if (aiText != null && !aiText.isBlank()) return aiText;
        return String.format("%s has thrown %d total strikes with %d landed (%.1f%% accuracy). " +
                "Their performance score of %.1f reflects career-long consistency in high-pressure exchanges.",
                name, thrown, landed, acc, f.getPerformanceScore());
    }

    // ── §5.12 PerformanceAnalyzer ─────────────────────────────────────────────

    public static double calculateEfficiencyScore(Fighter f) {
        int totalFights = f.getWins() + f.getLosses() + f.getDraws();
        double accuracy = f.getStrikesThrown() > 0
                ? (double) f.getStrikesLanded() / f.getStrikesThrown() * 100 : 0;
        double winRate = totalFights > 0 ? (double) f.getWins() / totalFights * 100 : 0;
        double streakBonus = Math.min(f.getWinStreak() * 5.0, 25);
        double score = accuracy * 0.4 + winRate * 0.4 + streakBonus;
        return round2(Math.min(score, 100));
    }

    public static String getMomentum(Fighter f) {
        if (f.getWinStreak() >= 3) return "SCORCHING";
        if (f.getWinStreak() >= 1) return "RISING";
        if (f.getLosses() > 0 && f.getWins() == 0) return "STRUGGLING";
        return "STABLE";
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private double clamp(double v, double min, double max) {
        return Math.max(min, Math.min(max, v));
    }

    private double round1(double v) {
        return Math.round(v * 10.0) / 10.0;
    }

    private static double round2(double v) {
        return Math.round(v * 100.0) / 100.0;
    }

    private String buildDynamicsPrompt(Fighter f1, Fighter f2, FightDynamicsResult h) {
        return String.format("Analyze fight dynamics: %s (ELO %.0f, %d-%d) vs %s (ELO %.0f, %d-%d). " +
                "Heuristic score: %.1f%% vs %.1f%%. Edge: %s. Provide a 2-sentence analysis.",
                h.f1Name(), f1.getEloRating(), f1.getWins(), f1.getLosses(),
                h.f2Name(), f2.getEloRating(), f2.getWins(), f2.getLosses(),
                h.f1Score(), h.f2Score(), h.edge());
    }
}
